pkg/certificate/mrc_reconciler_test.go

```diff
@@ -0,0 +1,456 @@
+		},
+		{
+			name: "mrc1 is passive and mrc2 is active",
+			mrcList: []*v1alpha2.MeshRootCertificate{
+				passiveMRC1,
+				activeMRC2,
+			},
+			expectedSigningMRC:    activeMRC2,
+			expectedValidatingMRC: passiveMRC1,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := tassert.New(t)
+
+			signingMRC, validatingMRC, err := getSigningAndValidatingMRCs(tt.mrcList)
+			if tt.expectedError != nil {
+				assert.ErrorIs(err, tt.expectedError)
+			} else {
+				assert.NoError(err)
+				assert.Equal(tt.expectedSigningMRC, signingMRC)
+				assert.Equal(tt.expectedValidatingMRC, validatingMRC)
+			}
+		})
+	}
+}
+
+func TestGetCertIssuers(t *testing.T) {
+	mrc1Name := "mrc1"
+	mrc2Name := "mrc2"
+	tests := []struct {
+		name                       string
+		signingMRC                 *v1alpha2.MeshRootCertificate
+		validatingMRC              *v1alpha2.MeshRootCertificate
+		expectedSigningIssuerID    string
+		expectedValidatingIssuerID string
+		currentSigningIssuerID     string
+		currentValidatingIssuerID  string
+		expectedError              error
+	}{
+		{
+			name:                       "same mrcs for signing and validating, issuer does not exist",
+			signingMRC:                 activeMRC1,
+			validatingMRC:              activeMRC1,
+			expectedSigningIssuerID:    mrc1Name,
+			expectedValidatingIssuerID: mrc1Name,
+		},
+		{
+			name:                       "same mrcs for signing and validating, issuer does exist as signingIssuer",
+			signingMRC:                 activeMRC1,
+			validatingMRC:              activeMRC1,
+			currentSigningIssuerID:     mrc1Name,
+			currentValidatingIssuerID:  mrc2Name,
+			expectedSigningIssuerID:    mrc1Name,
+			expectedValidatingIssuerID: mrc1Name,
+		},
+		{
+			name:                       "mrc1 is active and mrc2 is passive and issuer for mrc2 does not exist",
+			signingMRC:                 activeMRC1,
+			validatingMRC:              passiveMRC2,
+			currentSigningIssuerID:     mrc1Name,
+			currentValidatingIssuerID:  mrc1Name,
+			expectedSigningIssuerID:    mrc1Name,
+			expectedValidatingIssuerID: mrc2Name,
+		},
+		{
+			name:                       "mrc1 is active and mrc2 is passive and issuers exist",
+			signingMRC:                 activeMRC1,
+			validatingMRC:              passiveMRC2,
+			currentSigningIssuerID:     mrc2Name,
+			currentValidatingIssuerID:  mrc1Name,
+			expectedSigningIssuerID:    mrc1Name,
+			expectedValidatingIssuerID: mrc2Name,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := tassert.New(t)
+
+			m := &Manager{mrcClient: &fakeMRCClient{}}
+			if tt.currentSigningIssuerID != "" {
+				m.signingIssuer = &issuer{ID: tt.currentSigningIssuerID}
+			}
+			if tt.currentValidatingIssuerID != "" {
+				m.validatingIssuer = &issuer{ID: tt.currentValidatingIssuerID}
+			}
+
+			signingIssuer, validatingIssuer, err := m.getCertIssuers(tt.signingMRC, tt.validatingMRC)
```

Comment:
consider adding a check to see if underlying `m.mrcClient.GetCertIssuerForMRC(mrc)` is called by using the `fakemrcclient` to increment a number each time it is called.  A check like this would ensure that the cached versions are used when expected and we aren't making the additional calls when not expected.

Reply:
Great suggestion. Updated!